apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx

```diff
@@ -576,6 +586,58 @@ export class WorkshopForm extends React.Component {
     }
   }
 
+  renderFeeInput(validation) {
+    // If state.fee is null, there is no fee and no custom fee message.
+    // If state.fee is '', the user needs to provide a custom fee message.
+
+    const customizeFee = this.state.fee !== null;
+
+    return (
+      <FormGroup validationState={validation.style.fee}>
+        <ControlLabel>Fee</ControlLabel>
+
+        <div style={styles.noFeeContainer}>
+          <Radio
+            checked={!customizeFee}
+            inline
+            name="customize_fee"
+            value="no"
+            onChange={this.handleCustomizeFeeChange}
+            style={this.getInputStyle()}
+            disabled={this.props.readOnly}
+          >
+            No cost!
+          </Radio>
+        </div>
+
+        <div>
+          <Radio
+            checked={customizeFee}
+            inline
+            name="customize_fee"
+            value="yes"
+            onChange={this.handleCustomizeFeeChange}
+            style={{...this.getInputStyle(), ...styles.yesFeeRadio}}
+            disabled={this.props.readOnly}
+          >
+            <FormControl
+              type="text"
+              value={this.state.fee || ''}
+              id="fee"
+              name="fee"
+              onChange={this.handleFieldChange}
+              maxLength={30}
+              style={this.getInputStyle()}
+              disabled={this.props.readOnly || !customizeFee}
+              placeholder="Fee information"
+            />
+            <HelpBlock>{validation.help.fee}</HelpBlock>
+          </Radio>
+        </div>
+      </FormGroup>
+    );
+  }
+
```

Comment:
It's a shame to do so much custom work for a single entry on a single form.  It would have been nice to build this functionality into a form framework so we could trivially reuse it for any form.